brainstorming.md

```diff
@@ -0,0 +1,372 @@
+
+Where SOMA requires an explicit typing system, it utilizes the Arrow types and schema. SOMA has no specific requirements for the type or serialization system used by the underlying storage engine, other than it be capable of understanding and representing the Arrow types. It is understood and expected that each implementation of SOMA will have data type expressiveness limits (eg, just because you can express a type in he Arrow type system does not mean all SOMA implementations will understand it).
+
+In the following doc:
+
+- `primitive` types in this specification refers to Arrow primitive types, eg, `int32`, `float`, etc.
+- `string` refers to Arrow UTF-8 variable-length `string`, ie, `List<Char>`.
+- `simple` types include all primitive types, plus `string`.
+
+Other Arrow types are explicitly noted as such, eg, `Arrow.RecordBatch`.
+
+> ⚠️ **Issue** - is the use of Arrow as a base type language acceptable? There are others, up to and including inventing a new system, but it seemed prudent to adopt an existing, language neutral specification, which is already supported by many "engines".
+
+> ⚠️ **Issue** - are there parts of the Arrow type system that we wish to _explicitly exclude_ from SOMA? I have left this issue open (ie, no specific text) for now, thinking that we can come back and subset as we understand what complex types are required, and how much flexibility should be in this spec. We clearly need some complex types (eg, RecordBatch, List, etc) as they are implied by `string`, etc. My own preference would be to mandate a small set of primitive types, and leave the rest open to implementations to support as they feel useful.
+
+## Metadata (uns)
+
+All SOMA objects may be annotated with a small amount of metadata, in the form of a simple key/value map:
+
+- `uns` (map[string, simple]) - string-keyed mapping to any **simple** type (eg, int32, string, etc).
+
+Only `simple` types are supported in the metadata values. The metadata lifecycle is the same as its containing object, eg, it will be deleted when the containing object is deleted.
+
+> ⚠️ **Issue** - `uns` - do we need more complex typing for uns values, and if so, how to represent capture what can is required to be supported? (eg, anything Arrow can represent? a subset?)
+>
+> **Proposal**: restrict `uns` to a mapping of string keys and `simple` values. Extend to complex types in the future as we have concrete use cases which warrant the complexity.
+
+> ⚠️ **Issue** - the name `uns` is not really all that suggestive. **Proposal:** rename `uns` to `metadata`, eg, `print(soma.metadata['version'])`
```

Comment:
+1 -- this way neither seurat `misc` or anndata `uns` 'wins'; we have `neutral territory` in our naming

Reply:
@aaronwolen - this is a key-value store, ie, schema-free metadata. The only difference (currently) from `uns` is the requirement that all values in the store be primitives or strings (and it wouldn't surprise me if this evolves over time).

I personally prefer `metadata` and will integrate that change in the next rev.  However, I'm also very happy to change to something better if we find one....